pkg/apis/traceflow/v1/types.go

```diff
@@ -0,0 +1,158 @@
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   `json:"spec"`
+	Status `json:"status"`
+}
+
+// Spec describes the spec of the traceflow.
+type Spec struct {
+	Source      `json:"source"`
+	Destination `json:"destination"`
+	Packet      `json:"packet"`
+}
+
+// Source describes the source spec of the traceflow.
+type Source struct {
+	Namespace string `json:"namespace,omitempty"` // Namespace is the source namespace.
+	Pod       string `json:"pod,omitempty"`       // Pod is the source pod.
+}
+
+// Destination describes the destination spec of the traceflow.
+type Destination struct {
+	Namespace string `json:"namespace,omitempty"` // Namespace is the destination namespace.
+	Pod       string `json:"pod,omitempty"`       // Pod is the destination pod, exclusive with destination service.
+	Service   string `json:"service,omitempty"`   // Service is the destination service, exlusive with destination pod.
+	IP        string `json:"IP,omitempty"`        // IP is the destination IP.
+}
+
+// IPHeader describes spec of an IPv4 header. IPv6 not supported yet.
+type IPHeader struct {
+	SrcIP    string `json:"srcIP,omitempty"`    // SrcIP is the source IP.
+	Protocol int32  `json:"protocol,omitempty"` // Protocol is the IP protocol.
+	TTL      int32  `json:"ttl,omitempty"`      // TTL is the IP TTL.
+	Flags    int32  `json:"flags,omitempty"`    // Flags is the flags for IP.
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	ICMPEchoRequestHeader `json:"icmpEchoRequestHeader"`
+	UDPHeader             `json:"udpHeader"`
+	TCPHeader             `json:"tcpHeader"`
+}
+
+// ICMPEchoRequestHeader describes spec of an ICMP echo request header.
+type ICMPEchoRequestHeader struct {
+	ID       int32 `json:"id,omitempty"`       // ID is the ICMPEchoRequestHeader ID.
+	Sequence int32 `json:"sequence,omitempty"` // Sequence is the ICMPEchoRequestHeader sequence.
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	SrcPort int32 `json:"srcPort,omitempty"` // SrcPort is the source port.
+	DstPort int32 `json:"dstPort,omitempty"` // DstPort is the destination port.
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	SrcPort int32 `json:"srcPort,omitempty"` // SrcPort is the source port.
+	DstPort int32 `json:"dstPort,omitempty"` // DstPort is the destination port.
+	Flags   int32 `json:"flags,omitempty"`   // Flags are flags in the header.
+}
+
+// Packet includes payload info and headers info.
+type Packet struct {
+	PayloadString string `json:"payloadString,omitempty"` // PayloadString is the content in payload.
+	PayloadSize   int32  `json:"payloadSize,omitempty"`   // PayloadSize is the size of payload.
+
+	IPHeader        `json:"ipHeader"`
+	TransportHeader `json:"transportHeader"`
```

Comment:
I'm not sure whether they are intended to be defined as anonymous fields or not, as they have no field names but are non inline fields of json. 
I think it already caused problems when debugging, Let's make them clear here.
Either:
```
ipHeader IPHeader        `json:"ipHeader,omitempty"`
transportHeader TransportHeader `json:"transportHeader,omitempty"`
```
or
```
IPHeader        `json:",inline"`
TransportHeader `json:",inline"`
```

Reply:
Use former one.